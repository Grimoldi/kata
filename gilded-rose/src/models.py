AGED_BRIAR = "Aged Brie"
BACKSTAGE_PASS = "Backstage passes to a TAFKAL80ETC concert"
CONJURED_BREAD = "Conjured Bread"
SULFURAS = "Sulfuras, Hand of Ragnaros"
