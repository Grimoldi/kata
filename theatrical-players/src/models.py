from dataclasses import dataclass
from enum import StrEnum, auto


class PlayType(StrEnum):
    """Enumeration for play category."""

    COMEDY = auto()
    HISTORY = auto()
    PASTORAL = auto()
    TRAGEDY = auto()


class OutputType(StrEnum):
    """Enumeration for output type."""

    HTML = auto()
    PLAIN = auto()


@dataclass
class Performance:
    """Dataclass for a single performance."""

    _id: str
    name: str
    typology: PlayType


@dataclass
class Invoice:
    """Dataclass for a single invoice."""

    customer: str
    audience: int

    performance_id: str
