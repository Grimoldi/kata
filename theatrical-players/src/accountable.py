import importlib
import os
from importlib.util import module_from_spec, spec_from_file_location
from typing import Type

from accountables.core import PlayAccountable
from models import PlayType

ACCOUNTABLES: dict[PlayType, Type[PlayAccountable]] = dict()


def add_plugin(name: PlayType, accountable: Type[PlayAccountable]) -> None:
    """Adds a plugin."""
    ACCOUNTABLES[name] = accountable


def get_plugin(name: PlayType) -> Type[PlayAccountable]:
    """Gets a plugin by name."""
    return ACCOUNTABLES[name]


def plugin_exists(name: PlayType) -> bool:
    """Checks if a plugin exists."""
    return name in ACCOUNTABLES


def all_plugins() -> list[str]:
    """Gets all plugin names."""
    return list(ACCOUNTABLES.keys())


def import_module(name: str, path: str):
    """Imports a module from a file."""
    spec = spec_from_file_location(name, path)
    if spec:
        module = module_from_spec(spec)  # type: ignore
        spec.loader.exec_module(module)  # type: ignore
        return module
    raise ImportError(f"Could not import {name} from {path}")


def load_plugins_from_folder(folder: str) -> None:
    """Loads all modules from a folder."""
    for root, _, files in os.walk(folder):
        for file in files:
            if not file.endswith(".py") or file.startswith("core"):
                continue
            module_name = file[:-3]
            module_path = os.path.join(root, file)
            module = import_module(module_name, module_path)
            account_class = getattr(module, module_name.title())
            ACCOUNTABLES[module.get_play_accountable_name()] = account_class
