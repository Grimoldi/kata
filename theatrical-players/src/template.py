from typing import Protocol


def format_as_dollars(amount):
    return f"${amount:0,.2f}"


class Template(Protocol):
    """Protocol class for any output template."""

    def __init__(self) -> None:
        ...

    def add_customer(self, customer: str) -> None:
        """Add the customer to the billing."""
        ...

    def add_record(self, play_name: str, amount: float, audience: int) -> None:
        """Add a record to the template."""
        ...

    def add_total(self, amount: float) -> None:
        """Add the total amount to the template."""
        ...

    def add_credits(self, credit: float) -> None:
        """Add credits to the billing."""
        ...

    @property
    def bill(self) -> str:
        """Returns the bill."""
        ...


class RawTextTemplate:
    """Class to export the results as raw text."""

    def __init__(self) -> None:
        self._result = ""

    def add_customer(self, customer: str) -> None:
        """Add the customer to the billing."""
        self._result += f"Statement for {customer}\n"

    def add_record(self, play_name: str, amount: float, audience: int) -> None:
        """Add a record to the template."""
        self._result += (
            f" {play_name}: {format_as_dollars(amount/100)} ({audience} seats)\n"
        )

    def add_total(self, amount: float) -> None:
        """Add the total amount to the template."""
        self._result += f"Amount owed is {format_as_dollars(amount/100)}\n"

    def add_credits(self, credit: float) -> None:
        """Add credits to the billing."""
        self._result += f"You earned {credit} credits\n"

    @property
    def bill(self) -> str:
        """Returns the bill."""
        return self._result


class HtmlTemplate:
    """Class to export the results as HTML."""

    def __init__(self) -> None:
        self._customer = ""
        self._records = list()
        self._total = 0.0
        self._credit = 0.0

    def add_customer(self, customer: str) -> None:
        """Add the customer to the billing."""
        self._customer = customer

    def add_record(self, play_name: str, amount: float, audience: int) -> None:
        """Add a record to the template."""
        self._records.append((play_name, format_as_dollars(amount / 100), audience))

    def add_total(self, amount: float) -> None:
        """Add the total amount to the template."""
        self._total = amount

    def add_credits(self, credit: float) -> None:
        """Add credits to the billing."""
        self._credit = credit

    @property
    def bill(self) -> str:
        """Returns the bill."""
        content_rows = ""
        for record in self._records:
            content_rows += (
                "\n\t\t\t<tr>"
                f"<td>{record[0]}</td>"
                f"<td>{record[1]}</td>"
                f"<td>{record[2]} seats"
                "</tr>"
            )
        return (
            "<html>"
            "\n\t<style>"
            "\n\t\tth, td {"
            "\n\t\t\tpadding: 15px;"
            "\n\t\t}"
            "\n\t</style>"
            "\n\t<body>"
            "\n\t\t<table border=1>"
            f'\n\t\t\t<tr><td colspan=3><p align="right">{self._customer}</td></tr>'
            f"{content_rows}"
            f"\n\t\t\t<tr><td colspan=3>Amount owed is <font color=orange>{format_as_dollars(self._total/100)}</font>.</td></tr>"
            f"\n\t\t\t<tr><td colspan=3>You earned <font color=green>{self._credit} credits</font>.</td></tr>"
            "\n\t\t</table>"
            "\n\t</body>"
            "\n</html>"
        )
