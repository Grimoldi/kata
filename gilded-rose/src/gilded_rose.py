# -*- coding: utf-8 -*-

import models as m
import updater as u
from item import Item

updater_mapping = {
    m.AGED_BRIAR: u.AgedBrieUpdater(),
    m.BACKSTAGE_PASS: u.BackstagePassUpdater(),
    m.SULFURAS: u.SulfurasUpdater(),
    m.CONJURED: u.ConjuredUpdater(),
}


class GildedRose(object):
    def __init__(self, items):
        self.items = items

    def update_quality(self) -> None:
        for item in self.items:
            self._update_quality_of_single_item(item)

    def _update_quality_of_single_item(self, item: Item) -> None:
        updater = self._get_updater(item)
        updater.update_quality(item)
        updater.update_sellin(item)

    def _get_updater(self, item: Item) -> u.ItemUpdater:
        return updater_mapping.get(item.name, u.DefaultUpdater())
