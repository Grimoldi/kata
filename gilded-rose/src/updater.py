from typing import Protocol

import src.models as m
from src.item import Item


class ItemUpdater(Protocol):
    def update_sellin(self, item: Item) -> None:
        ...

    def update_quality(self, item: Item) -> None:
        ...


class DefaultUpdater:
    """
    All items have a SellIn value which denotes the number of days we have to sell the item
    All items have a Quality value which denotes how valuable the item is
    At the end of each day our system lowers both values for every item
    Once the sell by date has passed, Quality degrades twice as fast
    The Quality of an item is never negative
    """

    def update_sellin(self, item: Item) -> None:
        _decrease_sellin_date(item)

    def update_quality(self, item: Item) -> None:
        if item.sell_in > 0:
            amount = 1
        else:
            amount = 2
        _decrease_quality(item, amount)


class AgedBrieUpdater:
    """Aged Brie actually increases in Quality the older it gets"""

    def update_sellin(self, item: Item) -> None:
        _decrease_sellin_date(item)

    def update_quality(self, item: Item) -> None:
        _increase_quality(item)


class BackstagePassUpdater:
    """
    Backstage passes, like aged brie, increases in Quality as its SellIn value approaches;
    Quality increases by 2 when there are 10 days or less and by 3 when there are 5 days or less but
    Quality drops to 0 after the concert
    """

    def update_sellin(self, item: Item) -> None:
        _decrease_sellin_date(item)

    def update_quality(self, item: Item) -> None:
        if item.sell_in <= 0:
            _drop_quality(item)
        elif item.sell_in < 6:
            _increase_quality(item, 3)
        elif item.sell_in < 11:
            _increase_quality(item, 2)
        else:
            _increase_quality(item)


class ConjuredUpdater:
    """Conjured items degrade in Quality twice as fast as normal items"""

    def update_sellin(self, item: Item) -> None:
        _decrease_sellin_date(item)

    def update_quality(self, item: Item) -> None:
        if item.sell_in > 0:
            amount = 2
        else:
            amount = 4
        _decrease_quality(item, amount)


class SulfurasUpdater:
    """Sulfuras, being a legendary item, never has to be sold or decreases in Quality"""

    def update_sellin(self, item: Item) -> None:
        pass

    def update_quality(self, item: Item) -> None:
        pass


def _decrease_quality(item: Item, amount: int = 1) -> None:
    item.quality = max(m.MIN_QUALITY, item.quality - amount)


def _increase_quality(item: Item, amount: int = 1) -> None:
    item.quality = min(m.MAX_QUALITY, item.quality + amount)


def _decrease_sellin_date(item: Item) -> None:
    item.sell_in -= 1


def _drop_quality(item: Item) -> None:
    item.quality = m.MIN_QUALITY
