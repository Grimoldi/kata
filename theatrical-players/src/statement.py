import math
from typing import Any

import src.models as m


def statement(
    invoice: dict[str, Any],
    plays: dict[str, Any],
    output: m.OutputType = m.OutputType.PLAIN,
):
    total_amount = 0
    volume_credits = 0
    result = f'Statement for {invoice["customer"]}\n'

    def format_as_dollars(amount):
        return f"${amount:0,.2f}"

    for perf in invoice["performances"]:
        play = plays[perf["playID"]]
        audience = perf["audience"]
        play_type = play["type"]
        play_name = play["name"]

        this_amount = _get_play_amount(play_type, audience)
        total_amount += this_amount
        volume_credits += _get_credit(play_type, audience)

        # print line for this order
        result += (
            f" {play_name}: {format_as_dollars(this_amount/100)} ({audience} seats)\n"
        )

    result += f"Amount owed is {format_as_dollars(total_amount/100)}\n"
    result += f"You earned {volume_credits} credits\n"
    return result


def _get_play_amount(play_type: str, audience: int) -> float:
    """Gets the billed amount by play type."""
    amount_calculator = PLAY_MAPPING.get(play_type)
    if amount_calculator is None:
        raise ValueError(f"unknown type: {play_type}")

    return amount_calculator(audience)


def _get_comedy_amount(audience: int) -> float:
    this_amount = 300_00
    if audience > 20:
        this_amount += 100_00 + 5_00 * (audience - 20)
    this_amount += 3_00 * audience
    return this_amount


def _get_history_amount(audience: int) -> float:
    this_amount = 500_00
    if audience > 40:
        this_amount += 120_00 + 4_00 * (audience - 40)

    this_amount += 3_00 * audience
    return this_amount


def _get_pastoral_amount(audience: int) -> float:
    this_amount = 250_00
    if audience > 15:
        this_amount += 5_00 * (audience - 15)

    return this_amount


def _get_tragedy_amount(audience: int) -> float:
    this_amount = 400_00
    if audience > 30:
        this_amount += 10_00 * (audience - 30)

    return this_amount


def _get_credit(play_type: str, audience: int) -> float:
    """Add volume credits."""
    volume_credits = _get_volume_credits(audience)
    if play_type == "comedy":
        volume_credits += _get_comedy_credit_increment(audience)
    return volume_credits


def _get_volume_credits(audience: int) -> float:
    return max(audience - 30, 0)


def _get_comedy_credit_increment(audience: int) -> float:
    """Add extra credit for every ten comedy attendees."""
    return math.floor(audience / 5)


PLAY_MAPPING = {
    "comedy": _get_comedy_amount,
    "history": _get_history_amount,
    "pastoral": _get_pastoral_amount,
    "tragedy": _get_tragedy_amount,
}
