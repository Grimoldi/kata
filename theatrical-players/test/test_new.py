import json

import pytest
import src.models as m
from approval_utilities.utils import get_adjacent_file
from approvaltests import verify
from src.statement import statement


def test_statement_with_new_play_types():
    with open(get_adjacent_file("invoice_new_plays.json")) as f:
        invoice = json.loads(f.read())
    with open(get_adjacent_file("new_plays.json")) as f:
        plays = json.loads(f.read())
    verify(statement(invoice, plays))


def test_statement_with_new_play_types_as_html():
    with open(get_adjacent_file("invoice_new_plays.json")) as f:
        invoice = json.loads(f.read())
    with open(get_adjacent_file("new_plays.json")) as f:
        plays = json.loads(f.read())
    verify(statement(invoice, plays, m.OutputType.HTML))
