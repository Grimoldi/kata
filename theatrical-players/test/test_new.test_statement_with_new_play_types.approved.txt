Statement for BigCoII
 Henry V: $831.00 (53 seats)
 As You Like It: $450.00 (55 seats)
Amount owed is $1,281.00
You earned 48 credits
